### PR TITLE
glt - Adjust post method for `/api/courses/post` to prohibit adding primaries that have multiple secondaries

### LINKED ISSUE #17: Adjust post method for `/api/courses/post` to prohibit adding primaries that have multiple secondaries
# Overview

Adjust the `/api/courses/post` endpoint to handle adding courses with primaries and secondaries the same way UCSB GOLD does.

# Details

To understand this issue, you first need to understand what we mean by _primary_ and _secondary_.

All UCSB courses have a primary, with a section number that ends with `00`. This primary is typically the "lecture".

Some UCSB courses also have one or more discussion sections, that end in `01` ,`02`, etc.; these go along with a specific lecture. When courses have a discussion section (a "secondary"):

    * on GOLD you register for the discussion section (the secondary), not the primary

    * the primary is automatically added to your schedule

In this issue, we are going to change the POST /api/courses/post endpoint to work like GOLD:

First: when you try to add a course by enroll code, the system already looks up the enroll code already. Here is the code that does it,from the PSCourseController:

```

  String body = ucsbCurriculumService.getSection(enrollCd, checkPsId.getQuarter());

        if(body.equals("{\"error\": \"401: Unauthorized\"}") || body.equals("{\"error\": \"Enroll code doesn't exist in that quarter.\"}")){

            throw new BadEnrollCdException(enrollCd);

        }

```

The code then goes on to add the course as long as it exists for that quarter; if not, an exception is thrown.

We are then going to check the section number. There are a few different cases

**CASE 1: section number doesn't end in 00 **. If it does NOT end in `00`, that means that the enrollment code is for a "secondary", i.e. a discussion section. If the JSON already contains information about the lecture section (specifically, it's enrollCd), then the add may proceed... BUT we should ALSO add the lecture section to the personal schedule!

That is, this code (which note, has a poor choice of variable name: `courses` should be `course` since it is only one course:)

```java

 PSCourse courses = new PSCourse();

        courses.setUser(currentUser.getUser());

        courses.setEnrollCd(enrollCd);

        courses.setPsId(psId);

        PSCourse savedCourses = coursesRepository.save(courses);

        return savedCourses;

```

Should be replaced with _something like this_ (note that this is just top of head coding.. not guaranteed to be 100% correct)

```

        ArrayList<PSCourse> savedCourses = new ArrayList<>();



        PSCourse secondary = new PSCourse();

        secondary.setUser(currentUser.getUser());

        secondary.setEnrollCd(enrollCdSecondary);

        secondary.setPsId(psId);

        PSCourse savedSecondary = coursesRepository.save(secondary);

        savedCourses.add(savedSecondary);



        PSCourse primary = new PSCourse();

        primary.setUser(currentUser.getUser());

        primary.setEnrollCd(enrollCdPrimary);

        primary.setPsId(psId);

        PSCourse savedPrimary = coursesRepository.save(primary);

        savedCourses.add(savedPrimary);

      

        return savedCourses;

```

You will also need to change the return type from:

```

    public PSCourse postCourses(

```

to

```

    public List<PSCourse> postCourses(

```

**CASE 2: section number does end in 00, but there are no secondaries **. This is the case for most graduate courses, and some undergrad courses that have no separate section (e.g. ENGR 101, which has a 3 unit lecture and _no section_.) In this case, there is no special action to be taken: just add the primary only.

**CASE 3: section number ends in 00, and there ARE secondaries **. If there is more than one secondary, this is the case where we need to throw an exception. I suggest throw new IllegalArgumentException(enrollCd + " is for a course with sections; please add a specific section and the lecture will be automatically added");

Note that to support CASE 3, if it isn't already there from some other PR, you'll need to change this line in `ApiController.java':

FROM:

```java

  @ExceptionHandler({ EntityNotFoundException.class, BadEnrollCdException.class })

```

TO:

```java

  @ExceptionHandler({ EntityNotFoundException.class, BadEnrollCdException.class, IllegalArgumentException.class })

```

We'll need to do something similar for DELETE.. I'll put that in as a separate story.

### PR DESCRIPTION
**Pull request by eduglt**
# Overview

Update:  Successfully deployed on Heroku QA at f22-5pm-2-courses.

Closes #17 

In this PR, our goal is to change the backend endpoint `POST /api/courses/post` to match the functionality of GOLD.  GOLD distinguishes between "primary" sections such as lectures and "secondary" sections such as discussion sections.  For any enroll code the user requests to add, there are three cases:

1.  The enroll code is for a secondary.
2. The enroll code is for a primary without any secondaries.
3. The enroll code is for a primary with (multiple?) secondaries.

In each case, we take the corresponding action:

1. Add the secondary and its primary.  E.g. automatically add the lecture that goes with a discussion section.
2. Add the primary alone.  This is for classes without discussions, for example.
3. You are forbidden from signing up for a lecture without a discussion section.  We have to throw `IllegalArgumentException`.

# How to test

1. Checkout the branch `Geordie-postCourses`.
2. Build and run.
3. Login with OAuth.
4. Load subjects and create a personal schedule (remember your psId).
5. Go to Swagger and under `ps-course-controller` find the endpoint `POST /api/courses/post`.
image

6. Now try testing each of the three cases with some enroll codes. (You can find enroll codes fitting each case on the main page
image

# Guide to the code

In hindsight, it might have been better to split this into two issues, because in order to implement the originally requested feature, I added another feature to another part of the program.

`pom.xml` - Accidentally updated with some dependencies before I realized that the project already has a JSON library.  Do not look at some early commits because they may have used another library.

`ApiController.java` - Updated to include IllegalArgumentException handling.

`UCSBCurriculumService.java` - The existing `postCourses` method that we have to update uses a method `getSection` from this class, where `getSection` corresponds to a UCSB API endpoint.  The problem is that this endpoint doesn't provide the "related" sections to a given section's enroll code.  E.g., the returned JSON only has one section.  There is another endpoint in the UCSB API that *does* have this feature, so I created a new method `getAllSections` that gets data from that endpoint.  I think that this additional feature is justified because without it, finding the related sections would require us to scan through many irrelevant sections provided by the existing endpoints.  I also did not outright replace `getSection`'s implementation with a new one, because I don't know what consequences it would have on other code in the project.  Therefore, I introduced a new method called `getAllSections`.

`PSCourseController.java` - This contains the updated implementation of `postCourses` using the new `getAllSections`.  Note that when the output from the UCSB is unexpected/malformed, the behavior defaults to the old `postCourses`, and the enroll code is just treated as a primary. (We could also throw some exception maybe? Idk what kind.) Probably doesn't matter as long as the UCSB API is reliable.

`UCSBCurriculumServiceTests.java` - Almost a copy/paste of the existing `getSection` tests.  The tests ensure that the new method fails properly when given bad inputs.

`SectionFixtures` - I initially added three example fixtures corresponding to each of the three cases previously described.  Later, it turned out that I needed several more variations to get proper coverage given all the branches in the new feature.  Several of these variations are unusual, in that it doesn't make sense for the UCSB API to output them.

`PSCourseControllerTests.java` - There are several tests.  There are three corresponding to the main three cases and a few more corresponding to edge cases that probably cannot occur.

# Note for future corresponding delete endpoint
When we have to implement the corresponding delete endpoint that removes a primary and secondary together, you can probably reuse `getAllSections` to make your life easier.

### CODE REVIEW

**Comment by eduglt** [comment]
Unrelated to this PR's feature, but when I was working on `UCSBCurriculumService`, I noticed that several of the logging methods just log the same template URL over and over.  Seemed strange to me, was wondering if it is not desirable.

image

**Review by liamjet — APPROVED** [review]
LGTM

**Review by pconrad — CHANGES_REQUESTED** [review]
Suggesting a change that should fix the compilation problem.

**Comment by pconrad** [comment]
> Unrelated to this PR's feature, but when I was working on `UCSBCurriculumService`, I noticed that several of the logging methods just log the same template URL over and over. Seemed strange to me, was wondering if it is not desirable.
> 
> image

I've have to look at that in context, but I think what you are seeing may be normal behavior.  One of the things we log is all calls to the UCSB Curriculum API; and if one of the course update jobs is in an loop, you would see that call 
over and over again.

**Review by pconrad — APPROVED** [review]

File: src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java [hunk 1]
```diff
@@ -187,6 +189,51 @@ public String getSection(String enrollCode, String quarter) {
         log.info("json: {} contentType: {} statusCode: {}", retVal, contentType, statusCode);
         return retVal;
     }
+
+    public String getAllSections(String enrollCode, String quarter) {
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java:193]: The only thing I don't quite get here is: why is this called `getAllSections`?  Sure if it is looking up by enrollCd, it is going to get exactly one section, and one section only, right?

What is the difference between the contract of the SECTION_ENDPOINT (from v1 of the api) and ALL_SECTIONS_ENDPOINT (from v3 of the api?)

File: src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java [hunk 2]
```diff
@@ -60,6 +60,8 @@ public UCSBCurriculumService(RestTemplateBuilder restTemplateBuilder) {
 
     public static final String SECTION_ENDPOINT = "https://api.ucsb.edu/academics/curriculums/v1/classsection/{quarter}/{enrollcode}";
 
+    public static final String ALL_SECTIONS_ENDPOINT = "https://api.ucsb.edu/academics/curriculums/v3/classes/{quarter}/{enrollcode}";
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java:63]: What distinguishes SECTION_ENDPOINT from ALL_SECTIONS_ENDPOINT?  I'm confused.

Comments are not usually encouraged in this style of code--we prefer the code to be self-documenting through variable name choice.  However, this might be an exception, where a comment or two would be really helpful.

**Review by pconrad — CHANGES_REQUESTED** [review]
I've marked this a "request changes", but really it's just a question I want to understand better before signing off...

The choice of name for ALL_SECTIONS_ENDPOINT and the choice of name for the method `getAllSections` is a bit confusing for me.    Can we discuss?

File: src/main/java/edu/ucsb/cs156/courses/controllers/PSCourseController.java [hunk 3]
```diff
@@ -110,25 +115,61 @@ public PSCourse getCourseById(
...
-            @ApiParam("psId") @RequestParam Long psId) {
+            @ApiParam("psId") @RequestParam Long psId) throws JsonProcessingException {
         CurrentUser currentUser = getCurrentUser();
         log.info("currentUser={}", currentUser);
 
         PersonalSchedule checkPsId = personalScheduleRepository.findByIdAndUser(psId, currentUser.getUser())
         .orElseThrow(() -> new EntityNotFoundException(PersonalSchedule.class, psId));
 
-        String body = ucsbCurriculumService.getSection(enrollCd, checkPsId.getQuarter());
+        String body = ucsbCurriculumService.getAllSections(enrollCd, checkPsId.getQuarter());
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/courses/controllers/PSCourseController.java:127]: Ok, I get it now....

File: src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java (hunk 1 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java:193]: nvm.. I get it now. but I do think we should take from your excellent writeup in the PR description, and convert that into Javadoc comment.

File: src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java [hunk 4]
```diff
@@ -187,6 +189,51 @@ public String getSection(String enrollCode, String quarter) {
         log.info("json: {} contentType: {} statusCode: {}", retVal, contentType, statusCode);
         return retVal;
     }
+
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java:192]: I suggest adding:

```suggestion
/**
 * This method retrieves all of the sections related to a certain enroll code. For
 *  example if the enrollCode is for a discussion section, the lecture section
 *  and related discussion sections will also be returned.
 */
```

Comment (pconrad) [reply src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java:192]: And a similar comment in front of:

```
public String getSection(String enrollCode, String quarter) {
```

maybe something like:

```
 /**
  * This method gets only the section matching the enroll code and qtr
  */
```

**Review by pconrad — CHANGES_REQUESTED** [review]
Ok, I have specific suggestions to add some brief comments.. but otherwise, LGTM.. nicely done.

**Comment by eduglt** [comment]
Not sure what the CI failure is a symptom of, but it looks like several IllegalStateException errors are popping up. Guess my comment broke the whole project!? :smile:

**Comment by eduglt** [comment]
The checks pass when I run them manually through GitHub actions at least.  I don't know how the PR checks are different from that.

**Review by pconrad — APPROVED** [review]
LGTM!
